framework/core/image.h

```diff
@@ -119,8 +119,77 @@ struct ImageBuilder : public allocated::Builder<ImageBuilder, VkImageCreateInfo>
 		return *this;
 	}
 
+	ImageBuilder &with_compression(VkPhysicalDevice const             &gpu,
+	                               VkImageCompressionFlagsEXT          compression_flags,
+	                               VkImageCompressionFixedRateFlagsEXT fixed_rate_flags = VK_IMAGE_COMPRESSION_FIXED_RATE_NONE_EXT)
+	{
+		compression_control.flags = compression_flags;
+
+		// Do not support compression for multi-planar formats
+		compression_control.compressionControlPlaneCount = 1;
+
+		if (VK_IMAGE_COMPRESSION_DEFAULT_EXT != compression_flags)
+		{
+			if (VK_IMAGE_COMPRESSION_FIXED_RATE_DEFAULT_EXT == compression_flags ||
+			    VK_IMAGE_COMPRESSION_FIXED_RATE_EXPLICIT_EXT == compression_flags)
+			{
+				auto supported_compression_properties = vkb::query_supported_fixed_rate_compression(gpu, create_info);
+
+				if (VK_IMAGE_COMPRESSION_FIXED_RATE_EXPLICIT_EXT == supported_compression_properties.imageCompressionFlags)
+				{
+					if (VK_IMAGE_COMPRESSION_FIXED_RATE_EXPLICIT_EXT == compression_flags)
+					{
+						auto compression_fixed_rate_flags_vector = vkb::fixed_rate_compression_flags_to_vector(supported_compression_properties.imageCompressionFixedRateFlags);
+						if (compression_fixed_rate_flags_vector.empty())
+						{
+							LOGE("Image supports fixed-rate compression but support list is empty");
+						}
+						else if (fixed_rate_flags & supported_compression_properties.imageCompressionFixedRateFlags)
+						{
+							// Use requested flags if supported
+							fixed_rate_flags_array[0]           = fixed_rate_flags;
+							compression_control.pFixedRateFlags = fixed_rate_flags_array;
+						}
+						else
+						{
+							compression_control.compressionControlPlaneCount = 0;
+
+							LOGW("This image (format = {}) does not support the requested fixed-rate compression ({})",
+							     to_string(create_info.format),
+							     image_compression_fixed_rate_flags_to_string(fixed_rate_flags));
+							LOGW("Falling back to default fixed-rate compression (it supports {})",
+							     image_compression_fixed_rate_flags_to_string(supported_compression_properties.imageCompressionFixedRateFlags));
+
+							compression_control.flags = VK_IMAGE_COMPRESSION_FIXED_RATE_DEFAULT_EXT;
+						}
+					}
+				}
+				else
+				{
+					LOGW("This image does not support fixed-rate compression, falling back to default compression. Possible reasons:");
+					LOGW("\t- Its format may not be supported (format = {})", to_string(create_info.format));
+					if (create_info.usage & VK_IMAGE_USAGE_STORAGE_BIT)
+					{
+						LOGW("\t- It is a storage image (usage = {})", image_usage_to_string(create_info.usage));
+					}
+					if (create_info.samples > 1)
+					{
+						LOGW("\t- It is a multi-sampled image (sample count = {})", to_string(create_info.samples));
+					}
+
+					compression_control.flags = VK_IMAGE_COMPRESSION_DEFAULT_EXT;
+				}
+			}
+		}
+
```

Comment:
I would update the `create_info.pNext` chain here.  The local code knows that the `compression_control.pNext` is empty, so you could just do 

```c++
compression_control.pNext = create_info.pNext;
create_info.pNext = &compression_control;
```

But that assumes that no one will ever call `with_compression` more than once with the same builder (which isn't a great assumption).  Hence using a helper function is valuable.

In my C++ bindings examples I have the following:

```c++
template <typename T, typename N>
bool containsNext(const T& t, const N& n) {
    const vk::BaseInStructure* tp = &reinterpret_cast<const vk::BaseInStructure&>(t);
    const vk::BaseInStructure* np = &reinterpret_cast<const vk::BaseInStructure&>(n);
    while (tp != nullptr) {
        if (tp == np) {
            return true;
        }
        tp = tp->pNext;
    }
    return false;
}

template <typename T, typename N>
void injectNext(T& t, N& n) {
    vk::BaseOutStructure* tp = &reinterpret_cast<vk::BaseOutStructure&>(t);
    vk::BaseOutStructure* np = &reinterpret_cast<vk::BaseOutStructure&>(n);
    np->pNext = tp->pNext;
    tp->pNext = np;
}
```

If they're adapted to the C bindings then the logic would become 

```c++
if (!containsNext(create_info, compression_control)) 
{
  injectNext(create_info, compression_control)
}
```

Reply:
I like this solution, but I think it's out of the scope of this PR 😅 I'd prefer to leave if for a future refactor if/when we have more structures to consider?